### PR TITLE
Make crossentropy and binarycrossentropy numerically stable

### PR DESCRIPTION
**Pull request by Paethon**
Follows the implementation in Keras

The slightly complicated way of getting the epsilon is so that yhat does not have to contain Float

### CODE REVIEW

**Comment by Paethon** [comment]
Fixes #435

File: src/layers/stateless.jl [hunk 1]
```diff
@@ -4,8 +4,15 @@ using NNlib: logsoftmax, logσ
 
 mse(ŷ, y) = sum((ŷ .- y).^2)/length(y)
 
-function crossentropy(ŷ::AbstractVecOrMat, y::AbstractVecOrMat; weight = 1)
-  -sum(y .* log.(ŷ) .* weight) / size(y, 2)
+function crossentropy(ŷ::AbstractVecOrMat, y::AbstractVecOrMat;
+                      weight = 1, ϵ = eps(one(log(one(eltype(ŷ))))))
```
Comment (MikeInnes) [inline src/layers/stateless.jl]: Perhaps just `eps(float(eltype(y)))`?

File: src/layers/stateless.jl [hunk 2]
```diff
@@ -4,8 +4,15 @@ using NNlib: logsoftmax, logσ
...
+    if typeof(ϵ) == eltype(ŷ)
+        clamp!(ŷ, 0 + ϵ, 1 - ϵ)
```
Comment (MikeInnes) [inline src/layers/stateless.jl]: Better to always to the out-of-place version here, which you can fuse with the rest of the computation anyway. Otherwise it could be surprising to find the input modified.

File: src/layers/stateless.jl (hunk 1 above)
Comment (Paethon) [reply src/layers/stateless.jl]: OK, `log` uses `float()` anyway, so should be ok :)

**Comment by Paethon** [comment]
OK, I implemented the changes (thanks for the help)

I also had to add versions of `float` and `eps` for `TrackedReal`. Hope I did not mess that up :)

**Comment by boathit** [comment]
@MikeInnes I think adding such an epsilon to mitigate the numerical instability is deceptive and may lead to wrong loss. 

Considering a model whose output of the linear layer is `o = [-600, 102.5]` and the ground truth `y = [0.8, 0.2]`, if we use such an implementation to compute the loss what we will get is `28.83`; however, the true loss is `562.0`

```julia
function crossentropy_ϵ(ŷ::AbstractVecOrMat, y::AbstractVecOrMat;
                      weight = 1, ϵ = eps(float(eltype(ŷ))))
     -sum(y .* log.(clamp.(ŷ, 0 + ϵ, 1 - ϵ)) .* weight) / size(y, 2)
end

o = [-600, 102.5]
y = [0.8, 0.2]

crossentropy_ϵ(softmax(o), y) # 28.83

using Flux: logitcrossentropy
# true loss
logitcrossentropy(o, y) # 562.0
```

Even worse, for  all small `o[1]`  the  `crossentropy_ϵ()` will always return the same loss and this cannot correctly penalize the neuron whose output is incorrect 

```julia
o1 = [-700, 102.5]
o2 = [-1000, 102.5]
crossentropy_ϵ(softmax(o1), y) # 28.83
crossentropy_ϵ(softmax(o2), y) # 28.83

## true loss
logitcrossentropy(o1, y) # 642.0
logitcrossentropy(o2, y) # 882.0
```

A more reasonable way to define `crossentropy`, I think, is to change the interface of `crossentropy` to use the log-probabilities, i.e., `crossentropy(logŷ, y)` rather than `crossentropy(ŷ, y)`. In other words, we explicitly force the users to use `logsoftmax` instead of `softmax` as their output layer. Actually this is exactly what PyTorch does, see [KL Divergence Loss](https://pytorch.org/docs/stable/nn.html#kldivloss) and [NLL](https://pytorch.org/docs/stable/nn.html#nllloss).

**Comment by MikeInnes** [comment]
@boathit it’s a good point to raise, but does that kind of thing actually come up? AFAIK cross entropy is defined on probablility distributions, in which case we can reasonably leave this undefined for numbers outside 0-1. 

I do think we should make it easier to be numerically stable here, and I think #85 will be a great way to do that, since we can just rewrite `crossentropy(softmax(x),y)` however we want to then.

**Comment by boathit** [comment]
@MikeInnes In any case, it is desirable to compute the cross entropy through the logits. So I am also kind of into #85.  But to unify the interface, do we also need a lazy sigmoid function?

**Comment by ToucheSir** [comment]
All of these loss functions now use an appropriate epsilon (mostly via `epseltype`), so this can be closed.
